marshal/composite_test.go

```diff
@@ -0,0 +1,171 @@
+		t, compositeTest{
+			st: compositeChild{
+				Field2: 2,
+				CompositeParent: CompositeParent{
+					Field1:          1,
+					ForceSendFields: []string{"Field1"},
+				},
+				CompositeSecondParent: &CompositeSecondParent{
+					Field3:          3,
+					ForceSendFields: []string{"Field3"},
+				},
+			},
+			jsonString:     `{"field1":1, "field2":2, "field3":3}`,
+			matchClassic:   true,
+			matchUnmarshal: true,
+		},
+	)
+
+}
+
+func TestCompositeNil(t *testing.T) {
+	element := compositeChild{
+		Field2: 2,
+		CompositeParent: CompositeParent{
+			Field1:          1,
+			ForceSendFields: []string{"Field1"},
+		},
+	}
+	result := executeCompositeMarshalTest(
+		t, compositeTest{
+			st:           element,
+			jsonString:   `{"field1":1, "field2":2}`,
+			matchClassic: true,
+			//CompositeSecondParent will be present due to limitations of the unmarshalling
+			matchUnmarshal: false,
+		},
+	)
+	element.CompositeSecondParent = &CompositeSecondParent{}
+	assert.Equal(t, element, result)
+
+}
+
+func TestCompositeDefault(t *testing.T) {
+	executeCompositeMarshalTest(
+		t, compositeTest{
+			st: compositeChild{
+				Field2: 0,
+				CompositeParent: CompositeParent{
+					Field1:          0,
+					ForceSendFields: []string{"Field1"},
+				},
+				CompositeSecondParent: &CompositeSecondParent{
+					Field3:          0,
+					ForceSendFields: []string{"Field3"},
+				},
+			},
+			jsonString:     `{"field1":0, "field3":0}`,
+			matchClassic:   true,
+			matchUnmarshal: true,
+		},
+	)
+}
+
+type compositeTest struct {
+	st         compositeChild
+	jsonString string
+	// Compare marshal results with normal marshal
+	matchClassic bool
+	// Unmarshal may not match, since ForceSendFields will be populated during
+	// custom unmarshal process
+	matchUnmarshal bool
+}
+
+func executeCompositeMarshalTest(t *testing.T, tc compositeTest) compositeChild {
+	// Convert to JSON
+	res, err := json.Marshal(tc.st)
+	assert.NoError(t, err, "error while executing custom marshal")
+	compareJSON(t, tc.jsonString, string(res))
+
+	var reconstruct compositeChild
+	err = json.Unmarshal(res, &reconstruct)
+	assert.NoError(t, err, "error while unmarshaling")
+	if tc.matchUnmarshal {
+		assert.Equal(t, tc.st, reconstruct)
+	}
+
+	return reconstruct
+}
+
+type noSendFieldChild struct {
+	*CompositeSecondParent
+	Field2 int `json:"field2"`
+}
+
+func (s *noSendFieldChild) UnmarshalJSON(b []byte) error {
+	return Unmarshal(b, s)
+}
+
+func (s noSendFieldChild) MarshalJSON() ([]byte, error) {
+	return Marshal(s)
+}
+
+func TestNoSendFieldChild(t *testing.T) {
+	st := noSendFieldChild{
+		Field2: 2,
+	}
+
+	res, err := json.Marshal(st)
+	assert.NoError(t, err, "error while executing custom marshal")
+	compareJSON(t, `{"field2":2}`, string(res))
+
+	var reconstruct noSendFieldChild
+	err = json.Unmarshal(res, &reconstruct)
+	assert.NoError(t, err, "error while unmarshaling")
+	st.CompositeSecondParent = &CompositeSecondParent{}
+	assert.Equal(t, st, reconstruct)
+
+}
```

Comment:
Tests the same behaviour in a different struct. The struct is different in that `noSendFieldChild` has only one anonymous field (vs 2 in the compositeChild). Go works a bit different when there is only one anonymous field. It will surface ForceSendFields to the containing struct. So I added a test to ensure everything is still correct.